Tests/Test.cs

```diff
@@ -0,0 +1,327 @@
+using System.Collections.Immutable;
+using FluentAssertions;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Time.Testing;
+
+namespace Tests;
+
+using static ShoppingCartCommand;
+using static ShoppingCartEvent;
+using static ShoppingCart;
+
+// Types Aliases
+using ProductItems = ImmutableDictionary<string, int>;
+using ProductItem = (Guid ProductId, int Quantity);
+using PricedProductItem = (Guid ProductId, int Quantity, decimal Price);
+
+// Commands
+public abstract record ShoppingCartCommand
+{
+    public record AddProduct(
+        ProductItem ProductItem,
+        Guid? ClientId = null
+    ): ShoppingCartCommand;
+
+    public record RemoveProduct(
+        PricedProductItem ProductItem
+    ): ShoppingCartCommand;
+
+    public record Confirm(
+        Guid ClientId
+    ): ShoppingCartCommand;
+
+    public record Cancel: ShoppingCartCommand;
+
+    private ShoppingCartCommand() { }
+}
+
+public abstract record ShoppingCartEvent
+{
+    public record Opened(
+        Guid? ClientId,
+        DateTimeOffset OpenedAt
+    ): ShoppingCartEvent;
+
+    public record ProductAdded(
+        PricedProductItem ProductItem,
+        DateTimeOffset AddedAt
+    ): ShoppingCartEvent;
+
+    public record ProductRemoved(
+        PricedProductItem ProductItem,
+        DateTimeOffset RemovedAt
+    ): ShoppingCartEvent;
+
+    public record Confirmed(
+        Guid ClientId,
+        DateTimeOffset ConfirmedAt
+    ): ShoppingCartEvent;
+
+    public record Cancelled(
+        DateTimeOffset CanceledAt
+    ): ShoppingCartEvent;
+
+    private ShoppingCartEvent() { }
```

Comment:
is this needed? its abstract innit?

Reply:
It’s best we can get from C# not to allow derive from the shopping cart class besides those classes listed inside. Sealed would block all inheritance, this makes only nested classes to derive, all outside technically will be able to derive but the compiler will say no because they can’t use a private constructor. We cannot have good things in C#… ;)

Reply:
is there a C# wishlist item to have `sealed` permit nested types to derive?
that would enable closing the list and total matching
having an actual trackable item for this lets one advocate for the feature much better...

Reply:
I'm not aware of such.